File: androidApp/src/main/java/com/grup/android/LoggedInViewModel.kt
package com.grup.android

import androidx.lifecycle.viewModelScope
import com.grup.APIServer
import com.grup.models.User
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.runBlocking
import kotlinx.datetime.Clock

abstract class LoggedInViewModel : androidx.lifecycle.ViewModel() {
    companion object {
        private const val STOP_TIMEOUT_MILLIS: Long = 5000
        private var loggedInApiServer: APIServer? = null

        fun injectApiServer(apiServer: APIServer) {
            loggedInApiServer = apiServer
        }
    }

    protected val apiServer: APIServer
        get() = loggedInApiServer!!

    protected val userObject: User
        get() = apiServer.user

    protected suspend fun closeApiServer() {
        apiServer.logOut()
        loggedInApiServer = null
    }

    protected fun <T> Flow<T>.asState() =
        this.let { flow ->
            runBlocking { flow.first() }.let { initialValue ->
                flow.stateIn(
                    viewModelScope,
                    SharingStarted.WhileSubscribed(STOP_TIMEOUT_MILLIS),
                    initialValue
                )
            }
        }

    protected fun <T> Flow<List<T>>.asInitialEmptyState() =
        this.stateIn(
            viewModelScope,
            SharingStarted.WhileSubscribed(STOP_TIMEOUT_MILLIS),
            emptyList()
        )

    protected fun <T> Flow<T>.asNotification(initialValue: T) =
        this.stateIn(
            viewModelScope,
            SharingStarted.Eagerly,
            initialValue
        )
}

File: iosApp/iosApp/LoggedInViewModel.swift
//
//  LoggedInViewModel.swift
//  iosApp
//
//  Created by Justin Xu on 4/11/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class LoggedInViewModel: ObservableObject {
    private let STOP_TIMEOUT_MILLIS = 5000
    private static var _apiServer: APIServer? = nil

    static func injectApiServer(apiServer: APIServer) {
        _apiServer = apiServer
    }
    
    var apiServer: APIServer {
        get { LoggedInViewModel._apiServer! }
    }
    
    var userObject: User {
        get { apiServer.user }
    }
    
    func closeApiServer() async throws {
        try await LoggedInViewModel._apiServer?.logOut()
        LoggedInViewModel._apiServer = nil
    }
}
